LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDevice.cs

```diff
@@ -223,199 +232,229 @@ internal LoRaDevice(string devAddr, string devEUI, ILoRaDeviceClientConnectionMa
-                }
-                else if (twin.Properties.Desired.Contains(TwinProperty.CN470JoinChannel))
-                {
-                    DesiredCN470JoinChannel = GetTwinPropertyIntValue(twin.Properties.Desired[TwinProperty.CN470JoinChannel].Value);
-                }
+            if (twin.Properties.Desired.Contains(TwinProperty.SensorDecoder))
+                SensorDecoder = twin.Properties.Desired[TwinProperty.SensorDecoder].Value as string;
 
-                if (twin.Properties.Desired.Contains(TwinProperty.Supports32BitFCnt))
+            InitializeFrameCounters(twin);
+
+            if (twin.Properties.Desired.Contains(TwinProperty.DownlinkEnabled))
+            {
+                DownlinkEnabled = GetTwinPropertyBoolValue(twin.Properties.Desired[TwinProperty.DownlinkEnabled].Value);
+            }
+
+            if (twin.Properties.Desired.Contains(TwinProperty.PreferredWindow))
+            {
+                var preferredWindowTwinValue = GetTwinPropertyIntValue(twin.Properties.Desired[TwinProperty.PreferredWindow].Value);
+                if (preferredWindowTwinValue == Constants.ReceiveWindow2)
+                    PreferredWindow = preferredWindowTwinValue;
+            }
+
+            if (twin.Properties.Desired.Contains(TwinProperty.Deduplication))
+            {
+                var val = twin.Properties.Desired[TwinProperty.Deduplication].Value as string;
+                _ = Enum.TryParse<DeduplicationMode>(val, true, out var mode);
+                Deduplication = mode;
+            }
+
+            if (twin.Properties.Desired.Contains(TwinProperty.ClassType))
+            {
+                if (string.Equals("c", (string)twin.Properties.Desired[TwinProperty.ClassType], StringComparison.OrdinalIgnoreCase))
                 {
-                    Supports32BitFCnt = GetTwinPropertyBoolValue(twin.Properties.Desired[TwinProperty.Supports32BitFCnt].Value);
+                    ClassType = LoRaDeviceClassType.C;
                 }
+            }
 
-                if (twin.Properties.Desired.Contains(TwinProperty.KeepAliveTimeout))
+            if (twin.Properties.Reported.Contains(TwinProperty.PreferredGatewayID))
+            {
+                this.preferredGatewayID = new ChangeTrackingProperty<string>(TwinProperty.PreferredGatewayID, twin.Properties.Reported[TwinProperty.PreferredGatewayID].Value as string);
+            }
+
+            if (twin.Properties.Reported.Contains(TwinProperty.Region))
+            {
+                var regionValue = twin.Properties.Reported[TwinProperty.Region].Value as string;
+                if (Enum.TryParse<LoRaRegionType>(regionValue, true, out var loRaRegion))
                 {
-                    var value = GetTwinPropertyIntValue(twin.Properties.Desired[TwinProperty.KeepAliveTimeout].Value);
-                    if (value > 0)
+                    if (Enum.IsDefined(typeof(LoRaRegionType), loRaRegion))
```

Comment:
Is this if requirement? TryParse cannot be true and, at the same time, the outcome loraRegion is not a LoRaRegionType, no?

Reply:
Impossible. But that's not my change. :) I should have pointed out that this file is tricky to review, as I removed one level of indention for readability, but that caused the diff tool to mark a lot of existing code as being changed. Maybe if you ignore white spaces, it would be a bit better. Anyway, there are quite a few improvements we can do in the parsing logic of the LoRaDevice. We have an item open to fix dynamic type casting in there and we can consider some fixes there.